cmd/juju/common/controller.go

```diff
@@ -91,57 +86,73 @@ func WaitForAgentInitialisation(
 	}
 
 	ctx.Infof("Contacting Juju controller%s to verify accessibility...", addressInfo)
-	apiAttempts := 1
-	for attempt := attempts.Start(); attempt.Next(); apiAttempts++ {
-		err = tryAPI(c)
-		if err == nil {
-			msg := fmt.Sprintf("\nBootstrap complete, controller %q is now available", controllerName)
-			if isCAASController {
-				msg += fmt.Sprintf(" in namespace %q", caasprovider.DecideControllerNamespace(controllerName))
-			} else {
-				msg += fmt.Sprintf("\nController machines are in the %q model", bootstrap.ControllerModelName)
+	apiAttempts := 0
+	stop := make(chan struct{}, 1)
+	defer close(stop)
+	err = retry.Call(retry.CallArgs{
+		Clock:    clock.WallClock,
+		Attempts: bootstrapReadyPollCount,
+		Delay:    bootstrapReadyPollDelay,
+		Stop:     stop,
+		Func: func() error {
```

Comment:
In another PR, so we don't convolute this one, we can split this massive function call up into the different func arguments. `Func` to do the actual call, `IsFatalError` can be used to know when to stop and `NotifyFunc` can be used to get the number of apiAttempts instead of doing it manually.

Reply:
Have made a note and will submit soon :+1: